### ISSUE TITLE
Assess the situation with Framework's trailing slash matching changes

### ISSUE THREAD

**Issue by bclozel** [spring-projects/spring-boot#31563]
Right now [our build is failing](https://ci.spring.io/teams/spring-boot/pipelines/spring-boot-3.0.x/jobs/build/builds/820) because of recent changes in Framework's defaults for trailing slash match: https://github.com/spring-projects/spring-framework/issues/28552.

This issue should assess the situation and:
* fix the build
* reconsider an option for trailing slash matching for backwards compatibility (see #31327)
* check if there is a way to consider an option for the main application, and another one for the actuator web support

**Comment by vpavic** [comment]
Note that there's also #31547 which I assume this issue partly duplicates.

**Comment by wilkinsona** [comment]
We see the two as separate problems. Boot's been using PathPatternParser as the default for some time now. #31547 is about adapting to Framework now using it by default too. This issue's about adapting to a change in how trailing slashes are handled.

**Comment by wilkinsona** [comment]
We've discussed this today and we are happy with things the way they are. We are not going to introduce a property to enable a deprecated Spring Framework feature.

**Comment by vpavic** [comment]
Can you clarify what does _happy with things the way they are_ mean?

I'm mostly interested in the first point from your previous comment:

> Endpoint paths are configured using `/**`. This means that both `/actuator/endpoint` and `/actuator/endpoint/` work, irrespective of the trailing slash matching configuration. Do we want them to continue to behave this way?

Am I reading this right that the Actuator endpoints will continue to behave the same as if trailing slash matching was on?

**Comment by vpavic** [comment]
The `EndpointRequestMatcher` sort of assumes the use of Spring Security (which might not be the case) and doesn't really solve the request matching ambiguity concerns that were expressed in spring-projects/spring-framework#28552. If possible, I'd like to avoid such behavior by default in Actuator endpoints as well.

Also, this is now a bit surprising as it diverges from the new default behavior in the Framework.

**Comment by rstoyanchev** [comment]
`*` and `**` are defined as matching 0 or more characters, and likewise `{*rest}` matches 0 or more path segments, so it makes sense that `/actuator/endpoint/` matches when any of those are used.

A path variable on, the the other hand, expects a path segment, and so `/actuator/endpoint/{var}` does not match `/actuator/endpoint/` does not match, which means a combination of the following could do it:
```
/actuator/endpoint
/actuator/endpoint/{var}
/actuator/endpoint/{var}/**
```

That said, taking a step back, the trailing slash option essentially is about retrying with the same pattern + trailing slash. A pattern such as `/actuator/endpoint/**` is effectively a match by prefix, which expects `/actuator/endpoint` at a minimum, and it can be followed by anything, and that means anything, which to me includes a trailing slash as well. Furthermore, such the extra trailing slash check goes the end of the pattern, e.g. after the `**`, but it would never get there since `**` matches anything.

In short, I'm questioning whether a prefix pattern is in the same category as other patterns. I would expect a similar strategy for securing through a pattern by prefix and there should be no mismatch there as far as I can see.
